neps/nep-0539.md

```diff
@@ -0,0 +1,307 @@
+
+## Motivation
+
+We want to guarantee the Near Protocol blockchain operates stable even during
+congestion.
+
+Today, when users sends transactions at a higher rate than the network can
+process them, receipts accumulate without limits. This leads to unlimited memory
+consumption on chunk producer's and validator's machines. Furthermore, the delay
+for a transaction from when it is accepted to when it finishes execution becomes
+larger and larger, as the receipts need to queue behind those already in the
+system.
+
+First attempts to limit the memory consumption have been added without protocol
+changes. This is known as "local congestion control" and can be summarized in
+two rules:
+
+- Limit the transaction pool to 100 MB.
+  https://github.com/near/nearcore/pull/9172
+- Once we have accumulated more than 20k delayed receipts in a shard,
+  chunk-producers for that shard stop accepting more transactions.
+  https://github.com/near/nearcore/pull/9222
+
+But this does not put any limits on what shards will accept for other shards.
+For example, when a particular contract is popular, the contract's shard will
+eventually stop accepting new transactions but all other shards keep accepting
+more and more transactions that produce receipts for the popular contract.
+Therefore the number of delayed receipts keeps growing indefinitely.
+
+Cross-shard congestion control addresses this issue by stopping new transactions
+at the source and delaying receipt forwarding when the receiving shard has
+reached its memory limit.
+
+## Specification
+
+The proposal adds fields to chunks headers, it adds a trie columns, it changes
+the rules to select transactions, and it changes the chunk execution flow. The
+next four sections specifies each of these changes in more details.
+
+### Chunk header changes
+
+We change the chunk header to include congestion information.
+Specifically, we include two indicators.
+
+```rust
+ShardChunkHeaderInnerV3 {
+  // ...all field from ShardChunkHeaderInnerV2
+  incoming_congestion: u16
+  general_congestion: u16
+}
+```
+
+This adds 4 bytes to the chunk header, increasing it from 374 bytes to 378
+bytes in borsh representation. (Assuming no validator proposals included.)
+This in turn increases the block size by 4 bytes per shard.
+
+While the numbers are stored as `u16`,for the remainder of the NEP
+specification, `incoming_congestion` and `general_congestion` values are
+interpreted as numbers between 0 and 1, dividing the stored `u16` by `u16::MAX`
+(65535).
+
+### Changes to transaction selection
+
+Today, transactions are taken from the chunk producer's pool until `tx_limit` is
+reached, where `tx_limit` is computed as follow.
+
+```python
+# old
+tx_limit = 500 Tgas if len(delayed_receipts) < 20_000 else 0
+```
+
+We replace the formula for the transaction limit to depend on the
+`general_congestion` computed in the previous shard variable (between 0 and 1):
+
+```python
+# new
+MIN_GAS = 5 Tgas
```

Comment:
Should it be 5 Tgas or 0 Tgas?

Reply:
I think 5 Tgas is better but it's up for debate.

The idea here is that we allow a small amount of gas for accepting new transactions to other shard, if we know those other shards have `incoming_congestion = 0`. (see condition further below)

The thing is, if we allow it to go to zero, we allow complete starvation of a shard by other shards.
In practical terms, if my accounts happens to be on the same shard as HOT and everyone uses HOT, I can no longer do anything on any shard with my account, because I need to start at my own shard. Even if I only spend 2 Tgas on function call conversion on the congested shard and then 302 Tgas on a different shard which is currently idle.

Also, the harm of including 5 Tgas seems small since we still have 995 Tgas left for working on receipts to reduce the congestion.

Reply:
I see, makes sense.